bin/comment_on_pr

```diff
@@ -1,14 +1,143 @@
+#    # the existing comment by passing `--if-exist delete`.
+#    comment_on_pr --id comment-id [--if-exist update|delete] "This is a comment"
+#
+#    # Delete the comment associated with the given comment id by passing an empty comment
+#    comment_on_pr --id comment-id [--if-exist delete]
+#
+#    # Deprecated: Post a new comment to the PR
+#    comment_on_pr "This is a comment" <github-token>
+#
+# Please note the comment argument can be a path to a file containing the comment, or a markdown.
 
-if ! command -v gh &> /dev/null; then
-	brew install gh
+# Check dependencies and print their versions for diagnosis purposes
+cat <<EOF
+> jq --version
+$(jq --version)
+
+> curl --version
+$(curl --version)
+
+EOF
+
+if [[ ! "${BUILDKITE_PULL_REQUEST:-invalid}" =~ ^[0-9]+$ ]]; then
+    echo "Error: this tool can only be called from a Buildkite PR job"
+    exit 1
+fi
+
+# Default options
+opt_comment_id=""
+opt_if_exist="update"
+opt_github_token="${GITHUB_TOKEN:-}"
+
+# If there are two arguments and they don't start with an option, then we assume
+# this utility is called in the deprecated way: comment_on_pr <comment> <github-token>
+if [[ $# == 2 && "$1" != "--"* ]]; then
+    if [[ -f "$1" ]]; then
+        arg_pr_comment=$(cat "$1")
+    else
+        arg_pr_comment="$1"
+    fi
+
+    opt_github_token=$2
+else
+    while [[ "$#" -gt 0 ]]; do
+        case $1 in
+            --id) opt_comment_id="$2"; shift ;;
+            --if-exist) opt_if_exist="$2"; shift ;;
+            --github-token) opt_github_token="$2"; shift ;;
+            --*) echo "Unknown option: $1"; exit 1 ;;
+            *) break ;;
+        esac
+        shift
+    done
+
+    case "$#" in
+        0)
+            arg_pr_comment=""
+            ;;
+        1)
+            if [[ -f "$1" ]]; then
+                arg_pr_comment=$(cat "$1")
+            else
+                arg_pr_comment="$1"
+            fi
+            ;;
+        *) echo "Error: too many arguments"; exit 1 ;;
+    esac
+fi
+
+export GITHUB_TOKEN="$opt_github_token"
+
+github_user=$(basename "$(dirname "${BUILDKITE_PULL_REQUEST_REPO%.git}")")
+github_repo=$(basename "${BUILDKITE_PULL_REQUEST_REPO%.git}")
+issues_endpoint="repos/${github_user}/${github_repo}/issues"
+
+all_pr_comments=$(
+    github_api "$issues_endpoint/${BUILDKITE_PULL_REQUEST}/comments" --fail \
+    || (>&2 echo "Error: The GitHub token is invalid" && exit 1)
+)
+
+# Find the existing PR comment
+if [[ -n "$opt_comment_id" ]]; then
+    if ! [[ "$opt_comment_id" =~ ^[a-zA-Z0-9_-]+$ ]]; then
+        echo "Error: comment id can only contain alphanumeric characters, dashes and underscores"
+        exit 1
+    fi
+
+    # Do not change this variable value, it's used to identify the comments posted by this utility.
+    comment_body_id="<!-- DO NOT REMOVE ci-toolkit-comment-identifier: $opt_comment_id -->"
+    existing_comment_id="$(echo "$all_pr_comments" \
```

Comment:
🤔  I am not sure about this but I hope that if the `all_pr_comments` JSON contains special characters and `"…"` quotes in it, this `echo "$all_pr_comment" | jq …` won't risk getting tripped… worth testing with a concrete use case to ensure that this is not a problem

Reply:
Do you mean with a JSON like this?
```json
[{"body":"comment-id $#.% with \"quotes\"","id":1}]
```

I tested the above JSON using this shell and it worked as expected.
```
comment_body_id=comment-id; all_pr_comments='[{"body":"comment-id $#.% with \"quotes\"","id":1}]'; echo "$all_pr_comments" | jq ".[] | select(.body | contains(\"$comment_body_id\")) | .id"
```

Reply:
👌